autoload/clap/preview.vim

```diff
@@ -21,16 +21,29 @@ function! s:show_file_props(entry) abort
   call clap#preview#highlight_header()
 endfunction
 
+function! s:preview_lines_calc(h) abort
+  let l:layout = clap#layout#calc()
+  let l:remain = ( &lines - l:layout.row - l:layout.height - 1 ) / 2
+  let l:size = a:h
+  if l:remain <= 0
+    return 0
+  endif
+  if l:size <= l:remain
+    return l:size
+  endif
+  return l:remain
+endfunction
+
 if type(g:clap_preview_size) == v:t_number
   function! clap#preview#size_of(provider_id) abort
-    return g:clap_preview_size
+    return s:preview_lines_calc(g:clap_preview_size)
```

Comment:
Don't modify `clap#preview#size_of()` directly which is used to get the raw user defined setting, define a new function for that.

Reply:
I suggest tweaking the actual final preview size in `clap#floating_win#preview.show()`, https://github.com/liuchengxu/vim-clap/blob/2ad4ce34a143b7734577388fba765d637f64c8d5/autoload/clap/floating_win.vim#L289-L301

And then you don't have to call `clap#layout#calc()` always, use `s:display_opts` instead, https://github.com/liuchengxu/vim-clap/blob/2ad4ce34a143b7734577388fba765d637f64c8d5/autoload/clap/floating_win.vim#L57

Reply:
One issue I saw is preview.show take buffer as arguments. The plugin should load buffer according to the number of lines need to be displayed. in this case is the number of lines recalculated based on windows size. 
A extreme example, if user set the preview_size to 1,000,000. The code should trim the size before performa file buffer load. 
Also, if user set the number to an incorrect value, clap#preview#size_of should return the number that was used by clap instead of the value use set.

Reply:
> A extreme example, if user set the preview_size to 1,000,000. The code should trim the size before performa file buffer load.

A plugin can't prevent you from doing evil things.

Reply:
The code should detect the incorrect setup in the first place.  clap#preview#size_of may not be the function, but it is also too to do that when plugin need to create the float window.